### PR TITLE
[ESQL] copy ql disjunction rule into esql

### PR DESCRIPTION
**Pull request by not-napoleon**
Similar to https://github.com/elastic/elasticsearch/pull/105711, this migrates a QL optimization that depends on specific class types into ESQL to refer to the ESQL version of those classes.  This is necessary for https://github.com/elastic/elasticsearch/pull/105217, which changes the class hierarchy of the binary comparisons.

In this case, I didn't find any tests that exercise this optimization directly, although a failure in `PhysicalPlanOptimizerTests` alerted me to the problem.  To that end, I've added three tests to directly cover this optimization.

From the look of it, there's several more logical planner level optimizations that rely on the exact type of the comparison classes, so I'm going to migrate those as well.  For ease of review, I figured I'd do those in separate PRs.

### CODE REVIEW

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java [hunk 1]
```diff
@@ -1138,6 +1145,57 @@ protected In createIn(Expression key, List<Expression> values, ZoneId zoneId) {
...
+            Expression e = or;
+            // look only at equals and In
+            List<Expression> exps = splitOr(e);
+
+            Map<Expression, Set<Expression>> found = new LinkedHashMap<>();
+            ZoneId zoneId = null;
+            List<Expression> ors = new LinkedList<>();
+
+            for (Expression exp : exps) {
+                if (exp instanceof Equals eq) {
```
Comment (not-napoleon) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:1160]: Note, the key change here is that this now refers to esql's Equals, not the ql base Equals.

**Comment by not-napoleon** [comment]
Looking at it, there's a lot of tests in `OptimizerRulesTests` that we should port to ES|QL.  Notably, those tests should have caught the error that prompted this (from https://github.com/elastic/elasticsearch/pull/105217), but didn't because they don't actually exercise the ESQL classes.  Porting those tests seems like a good idea, and I'm open to doing it in this PR or opening an issue to do it as a follow up.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java (hunk 1 above)
Comment (costin) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:1160]: You could minimize the change by adding a templating method in the basic QL rule and overriding this in ESQL:
```
for (Expression exp : exps) {
   if (isEquals(exp)) {
     ....
   }
}

public boolean isEquals(Expression exp) {
        return exp instanceof QlEquals;
}

}

// rule in ESQL
public CombineDisjunctionsIn... extends ql.CombineDisjunctions {
      @Override
      protected boolean isEquals(Expression ex) {
             return ex instanceof EsqlEquals;
      }
}
```

**Review by costin — COMMENTED** [review]
If the change is small, this could be simplified by adding an overriding templating method.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java [hunk 2]
```diff
@@ -1118,28 +1117,6 @@ private static Project pushDownPastProject(UnaryPlan parent) {
...
-     * This rule looks for both simple equalities:
-     * 1. a == 1 OR a == 2 becomes a IN (1, 2)
-     * and combinations of In
-     * 2. a == 1 OR a IN (2) becomes a IN (1, 2)
-     * 3. a IN (1) OR a IN (2) becomes a IN (1, 2)
-     *
-     * This rule does NOT check for type compatibility as that phase has been
-     * already be verified in the analyzer.
-     */
-    public static class CombineDisjunctionsToIn extends OptimizerRules.CombineDisjunctionsToIn {
```
Comment (not-napoleon) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:1132]: Moved to `OptimizerRules` (ESQL version) to mirror the QL class hierarchy

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/OptimizerRules.java [hunk 3]
```diff
@@ -139,4 +154,80 @@ protected AttributeSet references(PhysicalPlan plan) {
...
+        protected In createIn(Expression key, List<Expression> values, ZoneId zoneId) {
+            return new In(key.source(), key, values);
+        }
+
+        protected Equals createEquals(Expression k, Set<Expression> v, ZoneId finalZoneId) {
+            return new Equals(k.source(), k, v.iterator().next(), finalZoneId);
+        }
+
+        @Override
+        protected Expression rule(Or or) {
```
Comment (not-napoleon) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/OptimizerRules.java:183]: Rule logic copied from QL version, but now pointing to ESQL versions of the classes

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java [hunk 4]
```diff
@@ -3441,15 +3475,6 @@ private void assertNullLiteral(Expression expression) {
         assertNull(expression.fold());
     }
 
-    // TODO: move these from org.elasticsearch.xpack.ql.optimizer.OptimizerRulesTests to org.elasticsearch.xpack.ql.TestUtils
```
Comment (not-napoleon) [inline x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java:3444]: Done.

**Review by costin — APPROVED** [review]
LGTM

> Porting those tests seems like a good idea, and I'm open to doing it in this PR or opening an issue to do it as a follow up.

👍 whatever you prefer.

**Comment by not-napoleon** [comment]
Actually, I did port the tests when I did the class hierarchy revision you asked for.  They're in the ESQL version of `OptimizerRuleTests`, which mirrors where they are in QL.

**Pull request merged**